modules/core/src/main/scala/ch/epfl/bluebrain/nexus/rdf/Encoder.scala

```diff
@@ -1,34 +1,94 @@
 package ch.epfl.bluebrain.nexus.rdf
 
+import java.time.{Instant, Period}
 import java.util.UUID
 
+import cats.{Contravariant, Foldable}
 import ch.epfl.bluebrain.nexus.rdf.Graph.{OptionalGraph, SetGraph}
 import ch.epfl.bluebrain.nexus.rdf.Iri.AbsoluteIri
 import ch.epfl.bluebrain.nexus.rdf.Node.{BNode, IriNode}
+import ch.epfl.bluebrain.nexus.rdf.Vocabulary.rdf
 
+import scala.annotation.tailrec
+import scala.concurrent.duration.{Duration, FiniteDuration}
+
+/**
+  * A type class that provides conversion from values of type ''A'' to an RDF [[Graph]]. Implementation inspired from
+  * the circe project (https://github.com/circe/circe).
+  */
 trait Encoder[A] extends Serializable { self =>
 
+  /**
+    * Converts a value to an RDF [[Graph]].
+    */
   def apply(a: A): Graph
 
+  /**
+    * Creates a new [[Encoder]] by applying a function to a value of type `B` to produce a value of type `A` and then
+    * encoding the result using this.
+    */
   final def contramap[B](f: B => A): Encoder[B] = new Encoder[B] {
     override def apply(a: B): Graph = self(f(a))
   }
-
 }
 
-object Encoder extends PrimitiveEncoderInstances with StandardEncoderInstances with RdfEncoderInstances {
+object Encoder
+    extends PrimitiveEncoderInstances
+    with StandardEncoderInstances
+    with RdfEncoderInstances
+    with LowPriorityEncoderInstances {
 
+  /**
+    * Summon an [[Encoder]] for the type `A` from the implicit scope.
+    */
   @inline
   final def apply[A](implicit instance: Encoder[A]): Encoder[A] = instance
 
+  /**
+    * Constructs an [[Encoder]] from a function.
+    */
   final def instance[A](f: A => Graph): Encoder[A] = new Encoder[A] {
     final def apply(a: A): Graph = f(a)
   }
+
+  implicit final val graphEncoderContravariant: Contravariant[Encoder] = new Contravariant[Encoder] {
+    override def contramap[A, B](fa: Encoder[A])(f: B => A): Encoder[B] = fa.contramap(f)
+  }
+}
+
```

Comment:
If you create this constructor here:
```scala
final def apply[C[_], A](fc: C[A] => Iterator[A])(implicit A: Encoder[A]): IterableAsListEncoder[C, A] = 
new IterableAsListEncoder[C, A](A) {
   override protected def toIterator(c: C[A]): Iterator[A] = fc(c)
}
```

You can use it to replace the code on `graphEncodeList `, `graphEncodeVector ` and `graphEncodeArray `

Reply:
done